### PR TITLE
Use pushFullExprCleanup for deferred destroy

### PR DESCRIPTION
**Pull request by usx95**
Instead of directly pushing the `Destroy`, we should use `pushFullExprCleanup` which handles conditional branches.
This fixes a backend crash due to 89ba7e183e6e2c64370ed1b963e54c06352211db.

Added minimized crash reproducer.

### CODE REVIEW

**Comment by usx95** [comment]
This fix is for a wrong refactoring change in https://github.com/llvm/llvm-project/commit/89ba7e183e6e2c64370ed1b963e54c06352211db and
I would be inclined towards landing this in a few hours to unblock the build breakage (if there are no review comments).
CC: @efriedma-quic

**Comment by cor3ntin** [comment]
@usx95 Could you maybe reduce a test from the crash? Thanks!

**Comment by usx95** [comment]
@cor3ntin Added tests for crash.

File: clang/test/CodeGenCXX/destory-with-init-list.cpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+class Object {
+public:
+  Object() = default;
+  struct KV;
+  explicit Object(std::initializer_list<KV> Properties);
+};
+
+class Value {
+public:
+  Value(std::initializer_list<Value> Elements);
```
Comment (efriedma-quic) [inline clang/test/CodeGenCXX/destory-with-init-list.cpp:25]: The use of initializer_list here doesn't seem necessary; you could remove it to reduce the size of the testcase.

Can we integrate this into an existing test file?

Comment (usx95) [reply clang/test/CodeGenCXX/destory-with-init-list.cpp]: Did you try this out ? I cannot reproduce the original crash without `std::initializer_list`.

Comment (efriedma-quic) [reply clang/test/CodeGenCXX/destory-with-init-list.cpp]:
```
class Object {
public:
  struct KV;
  Object(KV);
};

struct A { A(); ~A(); };
class Value {
public:
  Value(A);
  Value(const char *V);
  ~Value();
};

class ObjectKey {
public:
  ObjectKey(const char *S);
  ~ObjectKey();
};

struct Object::KV {
  ObjectKey K;
  Value V;
};
bool foo();
void DestroyInConditionalCleanup() {
  foo() ? Object{{"key1", A()}} : Object{{"key2", "val2"}};
}
```

Not really a big deal either way, though; just a minor cleanup.

**Review by efriedma-quic — APPROVED** [review]
LGTM

File: clang/test/CodeGenCXX/destory-with-init-list.cpp (hunk 1 above)
Comment (usx95) [reply clang/test/CodeGenCXX/destory-with-init-list.cpp]: Great. Thanks. I was able to reduce it further.

**Pull request merged**